text/2019-02-26-tikv-client-command-line.md

```diff
@@ -0,0 +1,232 @@
+
+## Detailed design
+
+The client will be created using the `tikv-client` crate and its already existing
+functionality. It will utilize `clap` to provide a simple wrapper around the
+crate API.
+
+### Install
+
+We will build this tool in the `tikv-cli` repository. It will be published.
+Users will be able to install it with `cargo install tikv-cli`. (See
+*Alternatives* for another option.)
+
+### Configuration
+
+The tool will retrieve its configuration from one of the following places, in
+order of precedence (greatest to least):
+
+* The command line options, such as `--pd 127.0.0.1:2379`.
+* A local configuration file `./.tikv-cli.toml`
+* The global configuration file `${XDG_CONFIG_HOME}/tikv-cli.toml`.
+* The default options (connect to localhost, no TLS).
+
+The options are:
+
+```toml
+[connection]
+# A list of `pd` endpoints.
+# Default: None
+pd_endpoints = [
+    "127.0.0.1:2379",
+    "127.0.0.2:2379",
+]
+
+[tls]
+# The path of the CA file for TLS.
+# Default: None
+ca_path = "./rsa.ca"
+# The path of the certificate file for TLS.
+# Default: None
+cert_path = "./rsa.cert"
+# The path of the key file for TLS.
+# Default: None
+key_path = "./rsa.key"
+
+[client]
+# Level of logging to use.
+# Available: critical, error, warning, info, debug, trace
+# Default: info
+logging = "info"
+# If the ouput JSON should be minified.
+# Default: false
+minify = false
+# The mode the client should be in.
+# All clients connected to a keyspace should use the same mode.
+# Available: raw, transaction
+# Default: transaction
+mode = "transaction"
+# If the tool should output on stderr the duration of commands.
+# Default: true
+output-durations = true
+```
+
+On the command line, these options have the same name, but with `_` replaced
+with `-` for idiomatic reasons.
+
+### Input/Output
+
+By default input is by positional arguments in `bash`/`sh` string format.
+
+By default the tool will output in prettyprinted JSON. For single entry results,
+it will return a string (including `"`), for key-value pairs it will use a JSON
+object `{ key: "", value: "" }`, for multiple results it will return a JSON
+array. Errors will use an `{ error: "" }` object. Opting out of prettyprint will
+be via the `--minify` flag.
+
+JSON was chosen because it is an efficient, pervasive, human-readable, modern
+standard that has many tools (eg `jq`) to use for post processing if needed.
+
+All standard (non-logging) output will be on `stdout`.
+
+It will feature configurable logging levels with `slog` which will output to `stderr`.
+
+The tool will exit zero on successful requests, and non-zero on unsuccessful
+requests. In the case of REPL mode, it shall exit zero unless the connection is
+found to be lost.
+
+### Interacting with it
+
+Users will interact with the tool through the command line. It will support both
+a one-off mode or a REPL (Read-eval-print-loop) style mode (See alternatives).

```

Comment:
Since it's the mode name, "a" is not needed here.
```suggestion
one-off mode and REPL (Read-eval-print-loop) style mode (See [Alternatives](#Alternatives)).
```